tfrecorder/schema.py

```diff
@@ -0,0 +1,179 @@
+    if v.type_name == "image_uri":
+      raw_schema['image_name'] = image_support_type
+      raw_schema['image'] = image_support_type
+      raw_schema['image_height'] = image_support_type
+      raw_schema['image_width'] = image_support_type
+      raw_schema['image_channels'] = image_support_type
+    else:
+      raw_schema[k] = schema_map[k]
+  return raw_schema
+
+
+def get_tft_coder(columns: List[str],
+                  schema_map: Dict[str, collections.namedtuple]
+                  ) -> tft.coders.CsvCoder:
+  """Gets a TFT CSV Coder.
+
+  Args:
+    columns: Ordered dataframe column names, from df.column.
+    schema_map: Schema map used to infer the schema.
+
+  Returns:
+    tft.coders.CsvCoder
+  """
+  feature_spec = dict()
+
+  # Because the DF column name order may not match the feature_spec order
+  # This maps existing column names to their feature spec (req part of
+  # namedtuple)
+  for col in columns:
+    feature_spec[col] = schema_map[col].feature_spec
+
+
+  metadata = dataset_metadata.DatasetMetadata(
+      schema_utils.schema_from_feature_spec(feature_spec))
+  return tft.coders.CsvCoder(columns,
+                             metadata.schema)
+
+
+def get_key(
+    type_name: str,
+    schema_map: Dict[str, collections.namedtuple]
+    ) -> Union[str, None]:
+  """Gets key of type 'type_name' from schema map if present,
+  otherwise returns None
+
+  Used by the beam pipeline to identify if images or split are present.
+  """
+  for k, v in schema_map.items():
+    if v.type_name == type_name:
+      return k
+  return None
+
+
+def get_raw_feature_spec(columns: List[str],
+                         schema_map: Dict[str, collections.namedtuple]
+                         ) -> Dict[str, tf.io.FixedLenFeature]:
+  """Gets RAW (pre TFT) feature spec."""
+
+  feature_spec = dict()
+
+  # Because the DF column name order may not match the feature_spec order
+  # this maps existing column names to their feature spec (req part of
+  # namedtuple)
+  for col in columns:
+    if schema_map[col].type_name == "image_uri":
+      # Modify feature_spec for extracted image, don't include image_uri.
+      # TODO(mikebernico) This only works in the case where the input has
+      # ONLY 1 image. Generalize to multiple images someday?
+      feature_spec['image_name'] = tf.io.FixedLenFeature([], tf.string)
+      feature_spec['image'] = tf.io.FixedLenFeature([], tf.string)
+      feature_spec['image_height'] = tf.io.FixedLenFeature([], tf.int64)
+      feature_spec['image_width'] = tf.io.FixedLenFeature([], tf.int64)
+      feature_spec['image_channels'] = tf.io.FixedLenFeature([], tf.int64)
+    else:
+      # Copy feature as-is.
+      feature_spec[col] = schema_map[col].feature_spec
+  return feature_spec
+
+
+def get_raw_metadata(columns: List[str],
+                     schema_map: Dict[str, collections.namedtuple]
+                     ) -> dataset_metadata.DatasetMetadata:
+  """Get's RAW (pre TFT) schema.
```

Comment:
nit: try """Returns metadata prior to TF Transform preprocessing"""

Reply:
Ack